File: LeetCode/Python/Easy/206-Reverse-Linked-List/solution.py
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None:
            return head
        
        new_head: ListNode=None

        current: ListNode=head

        next_node: ListNode=None

        while current!=None:
            next_node=current.next

            current.next=new_head

            new_head=current

            current=next_node

        return new_head

File: LeetCode/Rust/Easy/206-Reverse-Linked-List/solution.rs
impl Solution
{
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>>
    {
        if head.is_none()
        {
            return head;
        }

        let head_box_node: &Box<ListNode>=head.as_ref().unwrap();

        if head_box_node.next.is_none()
        {
            return head;
        }

        let mut new_head: Option<Box<ListNode>>=None;

        let mut next_node: Option<Box<ListNode>>=None;

        let mut current: Option<Box<ListNode>>=head;

        while current.is_some()
        {
            let mut pointer: Box<ListNode>=current.unwrap();
            
            next_node=pointer.next.take();

            pointer.next=new_head.take();

            new_head=Some(pointer);

            current=next_node;
        }

        new_head
    }
}